### ISSUE TITLE
False positive on dynamically created component in tests

### ISSUE THREAD

**Issue by MartijnHarte** [styled-components/styled-components#3410]
## System:
 - OS: macOS 11.2.1
 - CPU: (12) x64 Intel(R) Core(TM) i7-9750H CPU @ 2.60GHz
 - Memory: 521.50 MB / 16.00 GB
 - Shell: 5.8 - /bin/zsh
## Binaries:
 - Node: 10.22.0 - /usr/local/opt/nvm/versions/node/v10.22.0/bin/node
 - Yarn: 1.22.4 - /usr/local/bin/yarn
 - npm: 7.5.4 - /usr/local/opt/nvm/versions/node/v10.22.0/bin/npm
## npmPackages:
 - styled-components: ^5.2.1 => 5.2.1
 - jest: 24.9.0

## Reproduction

In my codebase I have globally mocked React with useRef so that I can properly test components that make use of the useRef hook.

I have found the following issue https://github.com/styled-components/jest-styled-components/issues/329#issuecomment-689701262 where ricardo-kagawa found that SC is calling useRef to determine whether or not the component is being created from within a component. This issue however is created on the `jest-styled-components` project, but even without using this package the warning is there.

It seems this check is also being run from within tests. This means that if useRef is mocked SC will think it's being created from within a component thus resulting in a lot of warnings when running tests.

## Steps to reproduce

The following test file shows a component that throws the warning: `The component styled.div with the id of "sc-bdfBwQ" has been created dynamically. You may see this warning because you've called styled inside another component. To resolve this only create new StyledComponents outside of any render method and function component.`

```
import styled from 'styled-components';

jest.mock('react', () => ({
    ...jest.requireActual('react'),
    useRef: jest.fn().mockReturnValue({ current: null }),
}));

test('it warns', () => {
    styled.div``;
});
```

## Expected Behavior

I expect to be able to mock useRef globally and not to have warnings.

## Actual Behavior

I mock useRef and get warnings of dynamically created components.

**Comment by kitten** [comment]
This is a usage issue. If you mock hooks (which I wouldn't really recommend doing, sorry 😅), you'll have to make sure that they at least throw outside of component contexts. If that's not really possible to you, you'll have to silence the warning in styled-components manually, which is _not triggered_ when `useRef` throws (or logs a specific error) as expected outside of components.

This may be more of a usage question, if you're looking at how to write these kinds of tests, with libraries like `react-testing-library` existing. Please refer with usage questions to Stack Overflow, as you're more likely to get quicker and helpful answers there ❤️ Repository issues on GitHub are great for that, issues, things that need to be resolved in the library itself, or can't be resolved by anyone but the maintainers. While I (and many others on other repos) are happy to eventually answer usage questions, styled-components is a busy repo and mostly stable and slow moving towards new releases, so you're not likely to get fast responses here
